File: python/train.py
"""
train.py - Training loop for the Heston PINN

Trains the PINN by minimizing a composite loss:
    L = lambda_pde * L_pde + lambda_bc * L_bc + lambda_ic * L_ic + lambda_data * L_data

Features:
    - Multi-scale collocation point generation
    - Adaptive loss weighting (gradient-based balancing)
    - Cosine annealing learning rate schedule
    - Curriculum learning (optional)
    - Validation against semi-analytical Heston prices
"""

import argparse
import time
import json
import numpy as np
import torch
import torch.optim as optim
from pathlib import Path
from typing import Dict, List, Optional, Tuple

from heston_pinn import (
    HestonPINN,
    HestonPINNResidual,
    generate_collocation_points,
    terminal_condition_loss,
    boundary_condition_loss,
    feller_boundary_loss,
)
from heston_analytical import heston_call_price


def compute_grad_norm(model: HestonPINN, loss: torch.Tensor) -> float:
    """Compute the norm of gradients of loss w.r.t. model parameters."""
    grads = torch.autograd.grad(
        loss, model.parameters(), retain_graph=True, allow_unused=True
    )
    total_norm = 0.0
    for g in grads:
        if g is not None:
            total_norm += g.norm().item() ** 2
    return np.sqrt(total_norm)


def validate_against_analytical(
    model: HestonPINN,
    heston_params: Dict[str, float],
    K: float,
    device: str = "cpu",
    n_points: int = 20,
) -> Tuple[float, float]:
    """
    Compare PINN prices against semi-analytical Heston prices.

    Returns:
        (mean_abs_error, max_abs_error)
    """
    S_vals = np.linspace(70, 130, n_points)
    v_vals = [0.02, 0.04, 0.08]
    t_val = 0.0  # Price at t=0

    errors = []
    for v0 in v_vals:
        for S in S_vals:
            # Analytical price
            analytical = heston_call_price(
                S, K, heston_params["T"], heston_params["r"],
                v0, heston_params["kappa"], heston_params["theta"],
                heston_params["sigma"], heston_params["rho"],
            )

            # PINN price
            S_t = torch.tensor([S], dtype=torch.float32, device=device)
            v_t = torch.tensor([v0], dtype=torch.float32, device=device)
            t_t = torch.tensor([t_val], dtype=torch.float32, device=device)

            with torch.no_grad():
                pinn_price = model(S_t, v_t, t_t).item()

            errors.append(abs(pinn_price - analytical))

    return np.mean(errors), np.max(errors)


def train(args):
    """Main training function."""
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print(f"Training on: {device}")

    # Heston parameters
    heston_params = {
        "kappa": args.kappa,
        "theta": args.theta,
        "sigma": args.sigma_vol,
        "rho": args.rho,
        "r": args.rate,
        "T": args.expiry,
    }

    # Check Feller condition
    feller = 2 * args.kappa * args.theta > args.sigma_vol ** 2
    print(f"Feller condition: {feller} "
          f"(2kT={2*args.kappa*args.theta:.4f} vs s^2={args.sigma_vol**2:.4f})")

    # Domain
    S_max = args.spot * 2.0
    v_max = max(args.theta * 5, 1.0)
    T = args.expiry
    K = args.strike

    # Create model
    model = HestonPINN(
        hidden_dim=args.hidden_dim,
        num_layers=args.num_layers,
        S_max=S_max,
        v_max=v_max,
        T_max=T,
    ).to(device)

    total_params = sum(p.numel() for p in model.parameters())
    print(f"Model parameters: {total_params:,}")

    # PDE residual computer
    residual = HestonPINNResidual(
        model,
        kappa=args.kappa,
        theta=args.theta,
        sigma=args.sigma_vol,
        rho=args.rho,
        r=args.rate,
    )

    # Optimizer and scheduler
    optimizer = optim.Adam(model.parameters(), lr=args.lr, weight_decay=1e-6)
    scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=args.epochs)

    # Loss weights (will be adapted)
    lambda_pde = 1.0
    lambda_bc = 10.0
    lambda_ic = 10.0
    lambda_feller = 5.0

    # Training history
    history = {
        "epoch": [],
        "loss": [],
        "pde_loss": [],
        "bc_loss": [],
        "ic_loss": [],
        "feller_loss": [],
        "val_mae": [],
        "val_max_err": [],
        "lr": [],
    }

    print("\n" + "=" * 80)
    print(f"{'Epoch':>7} | {'Loss':>10} | {'PDE':>10} | {'BC':>10} | "
          f"{'IC':>10} | {'Feller':>10} | {'Val MAE':>10} | {'LR':>10}")
    print("=" * 80)

    best_val_mae = float("inf")
    start_time = time.time()

    for epoch in range(args.epochs):
        optimizer.zero_grad()

        # Generate fresh collocation points each epoch
        S_col, v_col, t_col = generate_collocation_points(
            S_max, v_max, T, K, N_total=args.n_collocation, device=device
        )

        # 1. PDE residual loss
        L_pde = residual.pde_loss(S_col, v_col, t_col)

        # 2. Terminal condition loss
        L_ic = terminal_condition_loss(
            model, K, S_max, v_max, T,
            N_bc=args.n_boundary, option_type="call", device=device
        )

        # 3. Spatial boundary loss
        L_bc = boundary_condition_loss(
            model, K, args.rate, S_max, v_max, T,
            N_bc=args.n_boundary, option_type="call", device=device
        )

        # 4. Feller boundary loss (v=0)
        L_feller = feller_boundary_loss(
            model, args.kappa, args.theta, args.rate,
            S_max, T, N_bc=args.n_boundary, device=device
        )

        # Composite loss
        loss = (
            lambda_pde * L_pde
            + lambda_bc * L_bc
            + lambda_ic * L_ic
            + lambda_feller * L_feller
        )

        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
        optimizer.step()
        scheduler.step()

        # Adaptive weight adjustment every 200 epochs
        if epoch > 0 and epoch % 200 == 0:
            try:
                grad_pde = compute_grad_norm(model, lambda_pde * L_pde)
                grad_bc = compute_grad_norm(model, lambda_bc * L_bc)
                grad_ic = compute_grad_norm(model, lambda_ic * L_ic)

                mean_grad = (grad_pde + grad_bc + grad_ic) / 3.0
                if grad_bc > 1e-10:
                    lambda_bc = mean_grad / grad_bc
                if grad_ic > 1e-10:
                    lambda_ic = mean_grad / grad_ic
            except RuntimeError:
                pass  # Skip if gradient computation fails

        # Logging
        if epoch % args.log_every == 0:
            val_mae, val_max = validate_against_analytical(
                model, heston_params, K, device=device, n_points=10
            )

            current_lr = scheduler.get_last_lr()[0]

            history["epoch"].append(epoch)
            history["loss"].append(loss.item())
            history["pde_loss"].append(L_pde.item())
            history["bc_loss"].append(L_bc.item())
            history["ic_loss"].append(L_ic.item())
            history["feller_loss"].append(L_feller.item())
            history["val_mae"].append(val_mae)
            history["val_max_err"].append(val_max)
            history["lr"].append(current_lr)

            print(
                f"{epoch:7d} | {loss.item():10.6f} | {L_pde.item():10.6f} | "
                f"{L_bc.item():10.6f} | {L_ic.item():10.6f} | "
                f"{L_feller.item():10.6f} | {val_mae:10.6f} | {current_lr:10.2e}"
            )

            # Save best model
            if val_mae < best_val_mae:
                best_val_mae = val_mae
                save_path = Path(args.save_dir) / "best_model.pt"
                save_path.parent.mkdir(parents=True, exist_ok=True)
                torch.save({
                    "epoch": epoch,
                    "model_state_dict": model.state_dict(),
                    "optimizer_state_dict": optimizer.state_dict(),
                    "heston_params": heston_params,
                    "val_mae": val_mae,
                    "S_max": S_max,
                    "v_max": v_max,
                    "T_max": T,
                }, str(save_path))

    elapsed = time.time() - start_time
    print(f"\nTraining complete in {elapsed:.1f}s")
    print(f"Best validation MAE: {best_val_mae:.6f}")

    # Save final model
    final_path = Path(args.save_dir) / "final_model.pt"
    torch.save({
        "epoch": args.epochs,
        "model_state_dict": model.state_dict(),
        "heston_params": heston_params,
        "history": history,
        "S_max": S_max,
        "v_max": v_max,
        "T_max": T,
    }, str(final_path))

    # Save history
    history_path = Path(args.save_dir) / "training_history.json"
    # Convert numpy types to native Python for JSON serialization
    serializable_history = {}
    for k, v in history.items():
        serializable_history[k] = [float(x) if isinstance(x, (np.floating, float)) else int(x) for x in v]
    with open(history_path, "w") as f:
        json.dump(serializable_history, f, indent=2)

    print(f"Model saved to {final_path}")
    print(f"History saved to {history_path}")

    return model, history


def parse_args():
    parser = argparse.ArgumentParser(description="Train Heston PINN")

    # Heston parameters
    parser.add_argument("--kappa", type=float, default=2.0, help="Mean-reversion speed")
    parser.add_argument("--theta", type=float, default=0.04, help="Long-run variance")
    parser.add_argument("--sigma-vol", type=float, default=0.3, help="Vol of vol")
    parser.add_argument("--rho", type=float, default=-0.7, help="Correlation")
    parser.add_argument("--rate", type=float, default=0.05, help="Risk-free rate")

    # Option parameters
    parser.add_argument("--spot", type=float, default=100.0, help="Initial spot")
    parser.add_argument("--strike", type=float, default=100.0, help="Strike price")
    parser.add_argument("--expiry", type=float, default=1.0, help="Time to expiry")

    # Model architecture
    parser.add_argument("--hidden-dim", type=int, default=128, help="Hidden layer width")
    parser.add_argument("--num-layers", type=int, default=6, help="Number of residual blocks")

    # Training
    parser.add_argument("--epochs", type=int, default=5000, help="Training epochs")
    parser.add_argument("--lr", type=float, default=1e-3, help="Learning rate")
    parser.add_argument("--n-collocation", type=int, default=4096, help="Collocation points")
    parser.add_argument("--n-boundary", type=int, default=128, help="Boundary points")

    # Logging and saving
    parser.add_argument("--log-every", type=int, default=100, help="Log every N epochs")
    parser.add_argument("--save-dir", type=str, default="checkpoints", help="Save directory")

    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()
    model, history = train(args)

File: rust_pinn_heston/src/bin/train.rs
//! Training binary for the Heston PINN.
//!
//! Trains the neural network by minimizing the composite loss:
//!     L = lambda_pde * L_pde + lambda_bc * L_bc + lambda_ic * L_ic
//!
//! Uses finite-difference-based gradient estimation (SPSA) since we don't
//! have autograd in pure Rust. For production training, use the Python implementation
//! with PyTorch autograd.

use clap::Parser;
use pinn_heston::network::{generate_collocation_points, HestonPINN, TrainConfig};
use pinn_heston::analytical::heston_call_price;
use pinn_heston::HestonParams;
use rand::Rng;

#[derive(Parser, Debug)]
#[command(name = "train", about = "Train the Heston PINN model")]
struct Args {
    /// Number of training epochs
    #[arg(long, default_value_t = 1000)]
    epochs: usize,

    /// Hidden layer dimension
    #[arg(long, default_value_t = 64)]
    hidden_dim: usize,

    /// Number of hidden layers
    #[arg(long, default_value_t = 4)]
    num_layers: usize,

    /// Learning rate
    #[arg(long, default_value_t = 0.001)]
    lr: f64,

    /// Number of collocation points
    #[arg(long, default_value_t = 512)]
    n_collocation: usize,

    /// Spot price
    #[arg(long, default_value_t = 100.0)]
    spot: f64,

    /// Strike price
    #[arg(long, default_value_t = 100.0)]
    strike: f64,

    /// Time to expiry
    #[arg(long, default_value_t = 1.0)]
    expiry: f64,

    /// Output model path
    #[arg(long, default_value = "model.json")]
    output: String,

    /// Use crypto parameters
    #[arg(long)]
    crypto: bool,
}

/// Compute terminal condition loss: V(S, v, T) should equal payoff.
fn terminal_condition_loss(model: &HestonPINN, strike: f64, n_points: usize) -> f64 {
    let mut rng = rand::thread_rng();
    let mut total = 0.0;

    for _ in 0..n_points {
        let s = rng.gen_range(1.0..model.s_max);
        let v = rng.gen_range(0.001..model.v_max);
        let t = model.t_max; // At expiry

        let predicted = model.forward(s, v, t);
        let payoff = (s - strike).max(0.0); // Call payoff

        total += (predicted - payoff).powi(2);
    }

    total / n_points as f64
}

/// Compute boundary loss at S=0 (call should be 0).
fn boundary_loss_s0(model: &HestonPINN, n_points: usize) -> f64 {
    let mut rng = rand::thread_rng();
    let mut total = 0.0;

    for _ in 0..n_points {
        let v = rng.gen_range(0.001..model.v_max);
        let t = rng.gen_range(0.0..model.t_max);

        let predicted = model.forward(1e-6, v, t);
        total += predicted.powi(2); // Call at S=0 should be 0
    }

    total / n_points as f64
}

/// Compute PDE residual loss at collocation points.
fn pde_residual_loss(
    model: &HestonPINN,
    points: &[(f64, f64, f64)],
    params: &HestonParams,
) -> f64 {
    let mut total = 0.0;

    for &(s, v, t) in points {
        let residual = model.pde_residual(
            s, v, t,
            params.kappa, params.theta, params.sigma, params.rho, params.r,
        );
        total += residual.powi(2);
    }

    total / points.len() as f64
}

/// Compute total loss.
fn total_loss(
    model: &HestonPINN,
    points: &[(f64, f64, f64)],
    params: &HestonParams,
    strike: f64,
    config: &TrainConfig,
) -> f64 {
    let l_pde = pde_residual_loss(model, points, params);
    let l_ic = terminal_condition_loss(model, strike, config.n_boundary);
    let l_bc = boundary_loss_s0(model, config.n_boundary);

    config.lambda_pde * l_pde + config.lambda_ic * l_ic + config.lambda_bc * l_bc
}

/// Validate against analytical Heston prices.
fn validate(model: &HestonPINN, strike: f64, params: &HestonParams) -> (f64, f64) {
    let spots = [80.0, 90.0, 95.0, 100.0, 105.0, 110.0, 120.0];
    let v_vals = [0.02, 0.04, 0.08];

    let mut errors = Vec::new();

    for &v0 in &v_vals {
        let mut p = params.clone();
        p.v0 = v0;
        for &s in &spots {
            let analytical = heston_call_price(s, strike, params.r, &p);
            let pinn = model.forward(s, v0, 0.0);
            errors.push((analytical - pinn).abs());
        }
    }

    let mae = errors.iter().sum::<f64>() / errors.len() as f64;
    let max_err = errors.iter().cloned().fold(0.0_f64, f64::max);
    (mae, max_err)
}

fn main() {
    let args = Args::parse();

    let params = if args.crypto {
        HestonParams::crypto_default()
    } else {
        HestonParams::equity_default()
    };

    println!("=== Heston PINN Training ===");
    println!("Parameters: kappa={}, theta={}, sigma={}, rho={}, v0={}",
             params.kappa, params.theta, params.sigma, params.rho, params.v0);
    println!("Feller condition: {} (ratio={:.2})",
             params.feller_condition(), params.feller_ratio());
    println!("Architecture: {} hidden x {} layers", args.hidden_dim, args.num_layers);

    let s_max = args.spot * 2.0;
    let v_max = (params.theta * 5.0).max(1.0);
    let t_max = args.expiry;

    let mut model = HestonPINN::new(args.hidden_dim, args.num_layers, s_max, v_max, t_max);
    println!("Model parameters: {}", model.num_parameters());

    let config = TrainConfig {
        epochs: args.epochs,
        learning_rate: args.lr,
        n_collocation: args.n_collocation,
        n_boundary: 64,
        lambda_pde: 1.0,
        lambda_bc: 10.0,
        lambda_ic: 10.0,
        log_every: args.epochs / 20,
    };

    println!("\nTraining with SPSA gradient estimation...");
    println!("{:>7} | {:>10} | {:>10} | {:>10} | {:>10}",
             "Epoch", "Loss", "PDE", "Val MAE", "Val Max");
    println!("{}", "-".repeat(65));

    let perturbation = 0.01;
    let mut best_val_mae = f64::INFINITY;

    for epoch in 0..config.epochs {
        // Generate fresh collocation points
        let points = generate_collocation_points(
            s_max, v_max, t_max, args.strike, config.n_collocation,
        );

        // SPSA gradient estimation and update
        // Simultaneously Perturbation Stochastic Approximation
        let mut rng = rand::thread_rng();
        let c_k = perturbation / (1.0 + epoch as f64).powf(0.101);
        let a_k = args.lr / (1.0 + epoch as f64).powf(0.602);

        // For each layer, perturb all weights simultaneously
        for layer_idx in 0..model.layers.len() {
            let (rows, cols) = model.layers[layer_idx].weights.dim();

            // Random direction
            let delta_w: Vec<Vec<f64>> = (0..rows)
                .map(|_| {
                    (0..cols)
                        .map(|_| if rng.gen::<bool>() { 1.0 } else { -1.0 })
                        .collect()
                })
                .collect();

            let delta_b: Vec<f64> = (0..model.layers[layer_idx].biases.len())
                .map(|_| if rng.gen::<bool>() { 1.0 } else { -1.0 })
                .collect();

            // Perturb positive
            for i in 0..rows {
                for j in 0..cols {
                    model.layers[layer_idx].weights[[i, j]] += c_k * delta_w[i][j];
                }
            }
            for (i, d) in delta_b.iter().enumerate() {
                model.layers[layer_idx].biases[i] += c_k * d;
            }
            let loss_plus = total_loss(&model, &points, &params, args.strike, &config);

            // Perturb negative (2*c_k from positive)
            for i in 0..rows {
                for j in 0..cols {
                    model.layers[layer_idx].weights[[i, j]] -= 2.0 * c_k * delta_w[i][j];
                }
            }
            for (i, d) in delta_b.iter().enumerate() {
                model.layers[layer_idx].biases[i] -= 2.0 * c_k * d;
            }
            let loss_minus = total_loss(&model, &points, &params, args.strike, &config);

            // Restore and apply gradient
            let grad_est = (loss_plus - loss_minus) / (2.0 * c_k);

            for i in 0..rows {
                for j in 0..cols {
                    model.layers[layer_idx].weights[[i, j]] +=
                        c_k * delta_w[i][j] - a_k * grad_est * delta_w[i][j];
                }
            }
            for (i, d) in delta_b.iter().enumerate() {
                model.layers[layer_idx].biases[i] += c_k * d - a_k * grad_est * d;
            }
        }

        // Logging
        if config.log_every > 0 && epoch % config.log_every == 0 {
            let points_val = generate_collocation_points(
                s_max, v_max, t_max, args.strike, 128,
            );
            let current_loss = total_loss(&model, &points_val, &params, args.strike, &config);
            let pde_loss = pde_residual_loss(&model, &points_val, &params);
            let (val_mae, val_max) = validate(&model, args.strike, &params);

            println!("{:7} | {:10.6} | {:10.6} | {:10.6} | {:10.6}",
                     epoch, current_loss, pde_loss, val_mae, val_max);

            if val_mae < best_val_mae {
                best_val_mae = val_mae;
                if let Err(e) = model.save(&args.output) {
                    eprintln!("Failed to save model: {}", e);
                }
            }
        }
    }

    println!("\nTraining complete. Best validation MAE: {:.6}", best_val_mae);
    println!("Model saved to: {}", args.output);

    // Final validation
    println!("\nFinal price comparison:");
    let spots = [80.0, 90.0, 100.0, 110.0, 120.0];
    for &s in &spots {
        let pinn = model.forward(s, params.v0, 0.0);
        let analytical = heston_call_price(s, args.strike, args.expiry, &params);
        println!("  S={:.0}: PINN={:.4}, Analytical={:.4}, Error={:.4}",
                 s, pinn, analytical, (pinn - analytical).abs());
    }
}
